_posts/2016-09-04-build-maintainable-systems-with-hexagonal-architecture.md

```diff
@@ -0,0 +1,163 @@
+
+![Lottery system]({{ site.url }}{{ site.baseurl }}/assets/lottery.png)
+
+## Start from the core concepts
+
+We start the implementation from the system core. First we need to identify the core concepts of the lottery system. Probably the most important one is the lottery ticket. In lottery ticket you are supposed to pick the numbers and write your contact details to receive email notifications and possibly the prize money directly to your bank account. This leads us to write the following classes.
+
+<script src="http://gist-it.appspot.com/http://github.com/iluwatar/java-design-patterns/raw/master/hexagonal/src/main/java/com/iluwatar/hexagonal/domain/LotteryTicket.java?slice=24:71"></script>
+<script src="http://gist-it.appspot.com/http://github.com/iluwatar/java-design-patterns/raw/master/hexagonal/src/main/java/com/iluwatar/hexagonal/domain/LotteryNumbers.java?slice=30:121"></script>
+
+`LotteryTicket` contains `LotteryNumbers` and `PlayerDetails`. `LotteryNumbers` contains means to hold given numbers or generate random numbers and test the numbers for equality with another `LotteryNumbers` instance. [PlayerDetails](https://github.com/iluwatar/java-design-patterns/blob/master/hexagonal/src/main/java/com/iluwatar/hexagonal/domain/PlayerDetails.java) is a simple value object containing player's email address, bank account number and phone number.
+
+## The core business logic
+
+Now that we have the nouns presenting our core concepts we need to implement the core business logic that defines how the system works. In `LotterySystem` interface we write the methods that are needed by the lottery players and system administrators.
+
+<script src="http://gist-it.appspot.com/http://github.com/iluwatar/java-design-patterns/raw/master/hexagonal/src/main/java/com/iluwatar/hexagonal/domain/LotterySystem.java?slice=27:58"></script>
+
+The lottery players use `submitTicket()` to submit tickets for lottery round. After the draw has been performed `checkTicketForPrize()` tells the players whether they have won.
+
+For administrators `LotterySystem` has `resetLottery()` method for starting a new lottery round. At this stage the players submit their lottery tickets into the database and when the time is due the administration calls `performLottery()` to draw the winning numbers and check each of the tickets for winnings.
+
+<script src="http://gist-it.appspot.com/http://github.com/iluwatar/java-design-patterns/raw/master/hexagonal/src/main/java/com/iluwatar/hexagonal/domain/LotterySystemImpl.java?slice=34:107"></script>
+
+`LotterySystemImpl` has dependencies to lottery ticket database, banking and notifications ports. This implementation hardwires the adapters it uses. In a more sophisticated application we would use dependency injection to determine the implementation classes. The core logic is tested in [LotteryTest](https://github.com/iluwatar/java-design-patterns/blob/master/hexagonal/src/test/java/com/iluwatar/hexagonal/domain/LotteryTest.java).
```

Comment:
I can understand that the example needs to be kept easier to understand, but as Notifications, Database, and Banking are the ports of the system they ought to be injected into the system and not hardcode the dependencies.

The App (which is the production replica) should use actual implementations such as Email notifier, MongoRepository and such. The AppTest which will wire up the App for testing should use ConsoleNotifier, InMemoryRepository and so on. This will showcase the power of the pattern IMHO.